ledger/ledgercore/persistedacctdata.go

```diff
@@ -116,3 +117,458 @@ var EncodedMaxAppLocalStates = basics.EncodedMaxAppLocalStates
+func (e *ExtendedAssetHolding) ConvertToGroups(assets map[basics.AssetIndex]basics.AssetHolding) {
+	if len(assets) == 0 {
+		return
+	}
+
+	type asset struct {
+		aidx     basics.AssetIndex
+		holdings basics.AssetHolding
+	}
+	flatten := make([]asset, len(assets), len(assets))
+	i := 0
+	for k, v := range assets {
+		flatten[i] = asset{k, v}
+		i++
+	}
+	sort.SliceStable(flatten, func(i, j int) bool { return flatten[i].aidx < flatten[j].aidx })
+
+	numGroups := len(assets) / MaxHoldingGroupSize
+	if len(assets)%MaxHoldingGroupSize != 0 {
+		numGroups++
+	}
+	min := func(a, b int) int {
+		if a < b {
+			return a
+		}
+		return b
+	}
+
+	e.Count = uint32(len(assets))
+	e.Groups = make([]AssetsHoldingGroup, numGroups)
+	e.loaded = true
+
+	size := min(MaxHoldingGroupSize, len(assets))
+	currentGroup := 0
+	e.Groups[currentGroup] = AssetsHoldingGroup{
+		Count:         uint32(size),
+		MinAssetIndex: flatten[0].aidx,
+		groupData: AssetsHoldingGroupData{
+			AssetOffsets: make([]basics.AssetIndex, size, size),
+			Amounts:      make([]uint64, size, size),
+			Frozens:      make([]bool, size, size),
+		},
+		loaded: true,
+	}
+	prevAssetIndex := e.Groups[currentGroup].MinAssetIndex
+	prevGroupIndex := currentGroup
+	for i, a := range flatten {
+		currentGroup = i / MaxHoldingGroupSize
+		if currentGroup != prevGroupIndex {
+			e.Groups[prevGroupIndex].DeltaMaxAssetIndex = uint64(flatten[i-1].aidx - e.Groups[prevGroupIndex].MinAssetIndex)
+			size := min(MaxHoldingGroupSize, len(assets)-i)
+			e.Groups[currentGroup] = AssetsHoldingGroup{
+				Count:         uint32(size),
+				MinAssetIndex: flatten[i].aidx,
+				groupData: AssetsHoldingGroupData{
+					AssetOffsets: make([]basics.AssetIndex, size, size),
+					Amounts:      make([]uint64, size, size),
+					Frozens:      make([]bool, size, size),
+				},
+				loaded: true,
+			}
+			prevAssetIndex = e.Groups[currentGroup].MinAssetIndex
+			prevGroupIndex = currentGroup
+		}
+		e.Groups[currentGroup].groupData.AssetOffsets[i%MaxHoldingGroupSize] = a.aidx - prevAssetIndex
+		e.Groups[currentGroup].groupData.Amounts[i%MaxHoldingGroupSize] = a.holdings.Amount
+		e.Groups[currentGroup].groupData.Frozens[i%MaxHoldingGroupSize] = a.holdings.Frozen
+		prevAssetIndex = a.aidx
+	}
+	e.Groups[currentGroup].DeltaMaxAssetIndex = uint64(flatten[len(flatten)-1].aidx - e.Groups[currentGroup].MinAssetIndex)
+}
+
+// Loaded return a boolean flag indicated if groups are loaded or not
+func (e ExtendedAssetHolding) Loaded() bool {
+	return e.loaded
+}
+
+// SetLoaded sets loaded flag to true
+func (e *ExtendedAssetHolding) SetLoaded() {
+	e.loaded = true
+}
+
+// Clear removes all the groups, used in tests only
+func (e *ExtendedAssetHolding) Clear() {
+	e.loaded = false // ignored on serialization
+	for i := 0; i < len(e.Groups); i++ {
+		// ignored on serialization
+		e.Groups[i].groupData = AssetsHoldingGroupData{}
+		e.Groups[i].loaded = false
+	}
```

Comment:
instead of having this loop, use
```
e.Groups = []AssetsHoldingGroup{}
```
but more important, you missed setting the `e.Count` to zero.